### ISSUE TITLE
Race condition with generated ids.

### ISSUE THREAD

**Issue by binis2** [nemesis-software/nemesis-support#26]
Hello, 
We managed to hit a race condition when trying to import a lot of entities (customers) via rest. 
We tracked down the problem to PlatformIdentifierGenerator. We have multiple nemesis servers (7) that handle the requests. In that case the generator sometimes creates equal values.

After examination of the code it seems that you have somewhat foresaw the situation and added **clusterid** variable to mutate the generated id. However in the current code, **clusterid** is always 0. 

Can you please make a way so we can explicitly set **clusterid**? The ideal option will be via application.properties.

Best Regards,
Binis

**Comment by gochev** [comment]
@binis2 yes but this is not so easy to be made in cloud environment where the number of nodes can change very minute. 

Right now we advice that all the backend logic (I mean all REST calls and etc) should use a dedicated node for BACKEND processing (rest import/export/data get and etc).

In the initial architecture of the application this 7 nodes that were required ware ONLY for the frontend traffic. All load balancer stuff and etc should be ONLY for frontend. If backend is making calls for notifications, customer imports and etc they should use a dedicated node (not one of this 7) that will not create performance impact on the frontend for real users/customers.

**Comment by gochev** [comment]
Да обясня пак архитектурата е за 7 ноуда за фронтенд-а където не се импортват никакви потребители не се изпращат нотификации и има load balancer.

За бекенд-а ако искат да си dedicate-нат 1 от тия ноудове да си го викат него само и да падне и да не падне да не пречи на frontend customer-ите. Така няма да имате и този проблем.

**Comment by binis2** [comment]
Как като на клиента му "завалят" поръчки и почва да му фейлва създаването на ордери, например, ще обяснявам, че разбираш ли, трябва да се направят разни еквилибристики в load балансъра за да може да се поръчва само от един сървър. Или пък кат правя load тестове да им кажа да ги правят на един сървър или да се премирят с Х% грешки които има там като нещо нормално. Също така не виждам как броят на нодовете променя каквото и да било като всеки нод си има уникално ИД. Добавям нов нод с уникално ид всичко работи. 

Знаеш какъв е клиента и как като има проблем като захапва и си го гони до последно, независимо колко малко вероятно е нещо да се случи. Моля

            byte clusterid = 0;

да стане примерно

            public static byte clusterid = 0;

за да мога да си реша проблема по някакъв що годе нормален начин. Иначе ще се наложи да му бая на хибернейта и да го убеждавам, че "не е този генератора който търсиш"

**Comment by gochev** [comment]
Бинис ясна е работата @ptahchiev  го гледа.

Въпроса е че попринцип малко са нещата, които клиента СЪЗДАВА горе доло освен карта и ордър и някой и друг адрес няма друго, а тогава едва ли ще има особенно много колизии. 
Казвам докато се оправи или промени да ползвате само 1 ноуд за наливането на потребителите от Vivacom

**Comment by binis2** [comment]
Ако знаеше цялата история нямаше да даваш такива предложения :)

**Comment by gochev** [comment]
:D

**Comment by ptahchiev** [comment]
To reproduce this race condition then the follow must be met:
 - two of the same type of entities
 - in the same millisecond
 - are created with the same millisecond offset.

I thought this to be extremely unlikely to happen and that's why we wanted to add the clusterId (to be on the safe side). What is stopping me from exposing it as application property is the fact that it will not work when in a cloud environment and you don't know how many nodes are running at any given moment. 

I'm thinking of an alternative solution to add the `System.nanoTime` instead of clusterId to bring some more extra entropy. Will that be OK?

**Comment by gochev** [comment]
Otherwise if you have a better idea @binis2  you can always just patch the PlatformIdentifierGenerator class file.. I mean you can just replace it :) 

http://gochev.blogspot.bg/2014/07/patching-maven-library-with-your-custom.html

**Comment by binis2** [comment]
Well System.nanoTime should have been the pick from the start to be honest. Its more expensive in terms of CPU cycles but its more error prone. Should be a fine trade since its not expected for Nemesis to create too much entities anyway.

Still don't get it why you hesitate to expose clusterid. Adding removing node should not have effect if the node is configured properly and in worst case scenario it will have the same problem we have today. I don't insist to expose it as a property but still the user of the platform should have some control. Other satisfying approach will be for you to make a legal way to override the ID generator.

**Comment by ptahchiev** [comment]
Hi @binis2 I've included the following lines:

````
            byte numberOfNanoSecondsSinceTheStartOfTheLastSecond = (byte) Instant.now().getNano();
            longValue += ((long) numberOfNanoSecondsSinceTheStartOfTheLastSecond & 15L) << 44 & 263882790666240L;

````

instead of the cluster id and I have deployed a new snapshot version in our repository. Can you please give a proper stress testing and report if that fixes the problem?

To answer your question: we can't expose clusterID as application property because if I do so and deploy an application with clusterId=1 for example to a cloud environment, then it is possible that auto-scale kicks in and then starts more and more nodes with the same application property configuration:

[x] -- [x] --[x] ............ [x]

Where each of these nodes will have the same clusterId value, which actually defies the purpose of using it.

Let me know if the fix works. If not I will have to think for a new one over the weekend.
